### PR TITLE
[ROCm] Fix MIOpen RNN dropout mask reuse and gradient stall

### LINKED ISSUE #6339: [Issue]: Setting dropout via nn.GRU parameters causes unexpected behavior
### Problem Description

This was tested on both pytorch 2.12.0+rocm7.2 and pytorch 2.9.1+rocm7.2.1.gitff65f5bc.
A clean conda environment using torch 2.12+rocm7.2 was created for testing.

When building a character RNN from [example code](https://github.com/ageron/handson-mlp/blob/main/14_nlp_with_rnns_and_attention.ipynb) and with reference from previous training/validation benchmarks from an nvidia GPU, I encountered unexpected behavior when dropouts are introduced via parameters into nn.GRU. 

Specifically, when I call `self.gru = nn.GRU(embed_dim, hidden_dim, num_layers=n_layers,batch_first=True, dropout=dropout)` with 2 layers and dropout set to 0.1, training stalls on an AMD GPU, whereas training runs normally on an nvidia GPU. using nn.Dropout fixes the problem, which points to a bug in how dropout is implemented in nn.GRU (and possibly other similar classes).

loss stalls at ~3.07 irrespective of all hyperparameter tuning.
When manually placing nn.dropout between layers, normal loss optimization is recovered.

**Using dropouts w/ nn.gru (2 hidden layers, dropouts = .1) (ShakespeareModel, see example code):**
batch: 50, loss:3.6340441703796387
...
batch: 2300, loss:3.0771820545196533
...
batch: 5850, loss:3.067248821258545

**Using dropouts using nn.Dropout between 2 GRU layers (ShakespeareModel2 (!!), see example code):**
batch: 50, loss:3.6074438095092773
...
batch: 2300, loss:2.439793586730957
...
batch: 5850, loss:1.8970005512237549

**In the ROCm environment, switching to 'cpu' does not fix this problem.**

**I also ran this using a free T4 GPU in colab; nn.GRU dropouts worked normally and loss optimized normally (ShakespeareModel).**
batch: 50, loss:3.6132819652557373
...
batch: 2300, loss:2.5288307666778564
...
batch: 5850, loss:2.13324236869812

Efforts were made to resolve some non-determinism in the below script. I understand there are non-deterministic factors in training. Loss using nn.GRU dropouts does not minimize normally over multiple hyperparameter settings and seeds.

Thank you.

### Operating System

Ubuntu 24.04.4 LTS (Noble Numbat)

### CPU

CPU: AMD Ryzen 7 9700X 8-Core Processor

### GPU

Radeon RX 7900 XTX                 

### ROCm Version

7.2

### ROCm Component

_No response_

### Steps to Reproduce

**Loss should fall below 3 within 2000 batches**

```
from pathlib import Path
import urllib.request
import requests
import torch
import numpy as np
from torch.utils.data import DataLoader
import torch.nn as nn

SEED = 42
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.use_deterministic_algorithms(True)
device = 'cuda'


# Get shakespeare training set
url = "https://homl.info/shakespeare"
response = requests.get(url)
shakespeare_text = response.text

# Create encoder and decoder
vocab = sorted(set(shakespeare_text.lower()))
char_to_id = {char: index for index, char in enumerate(vocab)}
id_to_char = {index: char for index, char in enumerate(vocab)}

def encode_text(text):
    return torch.tensor([char_to_id[char] for char in text.lower()])

def decode_text(char_ids):
    return "".join([id_to_char[char_id.item()] for char_id in char_ids])


# Create Dataset
class CharDataset(torch.utils.data.Dataset):
  def __init__(self, text, window_size):
    self.text = text
    self.encoded_text = encode_text(self.text)
    self.window_size = window_size

  def __len__(self):
    # 1 short to preserve spot for target 
    return len(self.text) - self.window_size

  def __getitem__(self, idx):
    if idx >= len(self):
      raise IndexError("index outside valid range")
    end = idx + self.window_size
    window = self.encoded_text[idx:end]
    target = self.encoded_text[idx+1:end+1]
    return window, target

# Create dataloaders
window_length = 50
batch_size = 512
train_set = CharDataset(shakespeare_text[:1_000_000], window_length)
valid_set = CharDataset(shakespeare_text[1_000_000:1_060_000], window_length)
test_set = CharDataset(shakespeare_text[1_060_000:], window_length)
train_loader = DataLoader(train_set, batch_size=batch_size, shuffle=True, generator=torch.Generator().manual_seed(SEED), num_workers=0)
valid_loader = DataLoader(valid_set, batch_size=batch_size, num_workers=0)
test_loader = DataLoader(test_set, batch_size=batch_size, num_workers=0)

# Create Model
class ShakespeareModel(nn.Module):
    def __init__(self, vocab_size, n_layers=2, embed_dim=10, hidden_dim=128,
                 dropout=0.1):
        super().__init__()
        self.embed = nn.Embedding(vocab_size, embed_dim)
        self.gru = nn.GRU(embed_dim, hidden_dim, num_layers=n_layers,
                          batch_first=True, dropout=dropout)
        self.output = nn.Linear(hidden_dim, vocab_size)

    def forward(self, X):
        embeddings = self.embed(X)
        outputs, _states = self.gru(embeddings)
        return self.output(outputs).permute(0, 2, 1)

model = ShakespeareModel(len(vocab)).to(device)
opt = torch.optim.Adam(model.parameters())
loss_fn = nn.CrossEntropyLoss()
model.train()
cur_batch =0
for epoch in range(3):
  for i,(xb,yb) in enumerate(train_loader):
      cur_batch +=1
      xb = xb.to(device)
      yb = yb.to(device)
      opt.zero_grad()
      out = model(xb)
      loss = loss_fn(out.reshape(-1,out.size(1)), yb.reshape(-1))
      loss.backward(); opt.step()
      if cur_batch%50==0: print(f'batch: {cur_batch}, loss:{loss.item()}'
```

Here is the second model and basic training loop using nn.Dropout:
```

class ShakespeareModel2(nn.Module):
    def __init__(self, vocab_size, n_layers=2, embed_dim=10, hidden_dim=128,
                 dropout=0.1):
        super().__init__()
        self.embed = nn.Embedding(vocab_size, embed_dim)
        self.gru_1 = nn.GRU(embed_dim, hidden_dim, num_layers=1,
                          batch_first=True)
        self.dropout = nn.Dropout(dropout)
        self.gru_2 = nn.GRU(hidden_dim, hidden_dim, num_layers=1,
                          batch_first=True)
        self.output = nn.Linear(hidden_dim, vocab_size)

    def forward(self, X):
        embeddings = self.embed(X)
        outputs, _ = self.gru_1(embeddings)
        outputs = self.dropout(outputs)
        outputs, _ = self.gru_2(outputs)
        return self.output(outputs).permute(0, 2, 1)

model = ShakespeareModel2(len(vocab)).to(device)
opt = torch.optim.Adam(model.parameters())
loss_fn = nn.CrossEntropyLoss()
model.train()
cur_batch =0
for epoch in range(3):
  for i,(xb,yb) in enumerate(train_loader):
      cur_batch +=1
      xb = xb.to(device)
      yb = yb.to(device)
      opt.zero_grad()
      out = model(xb)
      loss = loss_fn(out.reshape(-1,out.size(1)), yb.reshape(-1))
      loss.backward(); opt.step()
      if cur_batch%50==0: print(f'batch: {cur_batch}, loss:{loss.item()}')
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by schung-amd**
# [ROCm] Fix MIOpen RNN dropout: gradient stall and missing per-step mask variation

Fixes [ROCm/ROCm#6339](https://github.com/ROCm/ROCm/issues/6339).

## Summary

Setting dropout via `nn.GRU(..., dropout=p)` with `num_layers>1` on ROCm causes training to stall, while the equivalent model using an explicit `nn.Dropout` between single-layer GRUs trains normally (and the fused form trains fine on NVIDIA). Reported in [ROCm/ROCm#6339](https://github.com/ROCm/ROCm/issues/6339): on the Shakespeare char-RNN reproducer, the fused `nn.GRU` dropout loss plateaus at **~3.07** and never improves across hyperparameters/seeds, whereas the `nn.Dropout` variant (and NVIDIA/Colab) reaches **~1.9-2.1**.

Root cause is in the MIOpen RNN dropout path (added in #144572). Two related defects, both fixed here at the PyTorch-wrapper level (no MIOpen change required):

1. **`weight_ih` gradient corruption -> the loss stall.** After the first step the forward built its dropout descriptor via `restore()` with `use_mask=true`, which makes MIOpen's forward dropout kernel *read* its mask from `reserveSpace` instead of generating one. MIOpen zero-initializes `reserveSpace` at the start of each forward-training call (`SetTensor(beta=0)` in `RNNForwardTrainingPackedTensors`), so the kernel reads an **all-zero mask -- i.e. an "all-drop" mask**. Backward reads the same all-zero mask, zeroing the gradients of the weights fed by the inter-layer dropout (`weight_ih`). The other weights (`weight_hh`, biases, embedding, output projection) still learn, so training makes early progress and then **stalls on a plateau** -- exactly the ~3.07 stall in ROCm/ROCm#6339.

2. **No per-step mask variation.** MIOpen's dropout kernel reads its xorwow PRNG state but never writes the advanced state back (the parameter is `const`; any writeback is dead-code-eliminated -- confirmed by disassembly). The wrapper seeded state once via `set()` then used `restore()` (no re-init), with the seed hardcoded to 0, so every training step would apply the *same* mask. The original PR explicitly deferred this (`// TODO: Implement seed setting for RNN dropout`, author noting *"this causes RNG machine to generate the same random numbers if we use the same seed... will later be addressed in a different PR."*). This is that follow-up.

### Fix

- **Per-step seed:** on each *training* forward, derive a fresh dropout seed from the default generator via `philox_engine_inputs` (host-side, no GPU sync, reproducible under `torch.manual_seed()`).
- **Re-seed each training forward:** `RNNDescriptors` runs `set()` (-> `InitPRNGState`) with the fresh seed on every training forward, giving each step an independent mask. **Backward does not re-seed** -- it replays the mask saved in `reserveSpace` during forward, keeping forward/backward consistent within a step.
- **Eval mode:** zero the dropout rate when `!train`, matching the cuDNN path (`cudnn/RNN.cpp`).
- **`use_mask=false` cleanup:** with reseeding, the forward no longer relies on reading a saved mask, so the `restore()` branch uses `use_mask=false`. Cleanup enabled by the reseed, not an independent fix (flipping it alone on the old code breaks `test_RNN_dropout_state`).

## Root cause detail

MIOpen cannot vary the mask across launches on its own: the dropout kernel loads PRNG state and computes the xorwow update, but emits no store of the advanced state (2 global stores total -- mask->reserveSpace and output->y; disassembly-confirmed). Mask variation must therefore be driven from the host via the seed feeding `rocrand_init` in `InitPRNGState` -- a wrapper concern.

## Test plan

- New regression test `test_RNN_dropout_gradients` (test/test_nn.py): asserts a multi-layer `nn.GRU(dropout>0)` produces a fresh mask each training step and non-zero `weight_ih` gradients. Verified to **fail on the original code** (`weight_ih` grad sum == 0.0) and **pass with this fix**.
- Reproducer from [ROCm/ROCm#6339](https://github.com/ROCm/ROCm/issues/6339): fused `nn.GRU(num_layers=2, dropout=0.1)` now trains normally -- loss falls past 3.0 within ~2000 batches and continues to **~2.08** by batch 5850, matching the `nn.Dropout`/NVIDIA reference (~1.9-2.1) instead of stalling at ~3.07.
- `test_RNN_dropout_state` -- passes, reliably (15/15 repeated runs).
- `test_RNN_dropout`, `test_RNN_change_dropout`, `test_RNN_cpu_vs_cudnn_with_dropout`, `test_RNN_cpu_vs_cudnn_no_dropout` -- pass.
- `test_cudnn_rnn` (fp16/bf16), `test_rnn_fused`, `test_rnn_retain_variables`, `test_{LSTM,GRU}_grad_and_gradgrad` -- pass.
- `test/nn/test_dropout.py` (14), `test/nn/test_packed_sequence.py` (13) -- pass.
- `test_modules.py` RNN/GRU/LSTM suite -- **264 tests pass** (18 skipped, 6 xfail; both baseline).
- Behavior checks: consecutive training forwards differ; eval deterministic; reproducible under `manual_seed`; `weight_ih` grad zero-fraction 0.000 across steps (vs 1.000 on original).
- gfx1201, ROCm 7.x.

## Files changed

- `aten/src/ATen/native/miopen/RNN_miopen.cpp` -- the fix
- `test/test_nn.py` -- `test_RNN_dropout_gradients` regression test

## Dropout state synchronization

Because every training forward now reseeds the PRNG buffer, the previously write-once shared buffer would be rewritten step-to-step, making unsynchronized sharing across models, streams, and devices a real hazard. The dropout state therefore mirrors the cuDNN path: one lazily allocated buffer per device in a process-wide cache, a host mutex for cross-thread exclusion, and a CUDAEvent ordering buffer uses across streams, held from descriptor setup through kernel enqueue.

Measured on gfx90a / HIP 7.14: the per-forward reseed costs ~220 us (PRNG init over the 640 KB xorwow state; Restore is ~0). Note: training-mode RNNs with dropout now touch the default CUDA generator (`philox_engine_inputs`), which asserts when HIP graph capture is active -- a host-side seed cannot be derived during capture, so this is a hard error rather than silent misbehavior.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by jeffdaily** [comment]
Excellent root-cause work -- the analysis checks out down to MIOpen internals (I verified the use_mask semantics and the backward reserveSpace replay against the MIOpen source, and reproduced the zeroed weight_ih gradients on gfx90a). I've pushed two maintainer commits and extended the PR body:

1. A merge of `viable/strict` (the branch predates the ROCm 7.14 upgrade and `_miopen_clear_dropout_state`).

2. A rework of the dropout-state storage to mirror cuDNN's `DropoutState` (per-device cache + mutex + event ordering). With your per-forward reseed, the old shared thread-local buffer became a rewrite-every-step hazard across models/streams/devices in a thread; the event machinery is the same solution cuDNN uses for exactly this. Also: the regression test now compares three consecutive forwards pairwise (on the broken code out1 != out2 held because the first forward generated a real mask -- only out2 == out3 exposed the repeated mask), plus small cleanups (stale TODO, unused includes).

Validated on gfx90a/HIP 7.14: all four TestNN RNN dropout tests pass, the #6339 repro trains with nonzero weight_ih gradients, and the reseed costs ~220 us/forward (measured with a standalone MIOpen probe; details in the body).

*Maintainer commits and this comment prepared with the assistance of an AI tool and reviewed before posting.*

**Review by jeffdaily — APPROVED** [review]
